### PR TITLE
HPCC-17483 Reduce contention on Roxie lazy IFileIO

### PR DESCRIPTION
**Pull request by jakesmith**
Reduce time that roxy lazy IO files hold a critical section
that is there to protect interaction with the caching mechanism.
Link active IFileIO instead and release CS.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Regression suite ran/passed.
Ran index based performance suite tests on Roxie.

### CODE REVIEW

File: roxie/ccd/ccdfile.cpp [hunk 1]
```diff
@@ -190,7 +190,8 @@ class CRoxieLazyFileIO : implements ILazyFileIO, implements IDelayedFile, public
     {
         try
         {
-            if (current.get()!=&failure)
+            if (current.get()==&failure)
+                return;
             {
```
Comment (richardkchapman) [inline roxie/ccd/ccdfile.cpp]: Remove the curlies on this block (or reverse the sense of the test and move the current.set into the block)?

File: roxie/ccd/ccdfile.cpp [hunk 2]
```diff
@@ -312,13 +321,14 @@ class CRoxieLazyFileIO : implements ILazyFileIO, implements IDelayedFile, public
 
     virtual size32_t read(offset_t pos, size32_t len, void * data) 
     {
-        CriticalBlock b(crit);
+        unsigned activeIdx;
+        Owned<IFileIO> active = getCheckOpen(activeIdx); // this object would close on last release
```
Comment (richardkchapman) [inline roxie/ccd/ccdfile.cpp]: Not sure that the comment here is helpful (or accurate)

**Comment by jakesmith** [comment]
@richardkchapman - do you have a good way of stress testing this?

**Comment by richardkchapman** [comment]
Not really

**Comment by richardkchapman** [comment]
@jakesmith Perf suite and/or regression suite is as good a way as any I suspect.

**Comment by jakesmith** [comment]
Yes, I'll run those and check.

**Comment by jakesmith** [comment]
@richardkchapman - please review

**Comment by richardkchapman** [comment]
Did the perf suite show any changes?

**Comment by jakesmith** [comment]
Inconclusive, the index queries took 144seconds +/- 4 secs with or without this change.
I don't think the perf. tests cause the kind of contention needed to show a gain here.

**Comment by richardkchapman** [comment]
Looks good to me. @ghalliday Can you sanity check too

**Pull request merged**